### DISCUSSION TITLE
method-signature-style with overloads, what's the recommended practice?

### DISCUSSION THREAD

**Discussion by yf-yang** [typescript-eslint/typescript-eslint#6957] [Technical Discussions]
Related: https://github.com/typescript-eslint/typescript-eslint/issues/2706, https://github.com/typescript-eslint/typescript-eslint/issues/4106

I enabled `method-signature-style`, here is an example 
``` typescript
interface testInterface {
  // first style
  // foo(a: number): number;
  // foo(a: string): string;
  // foo(a: number | string): number | string;

  // second style
  // foo: ((a: number) => number) & ((a: string) => string) & ((a: number|string) => number|string);

  // third style
  // foo: {
  //     (a: number): number;
  //     (a: string): string;
  //     (a: number | string): number | string;
  // }
}

class TestClass implements testInterface {
  foo(a: number): number;
  foo(a: string): string;
  foo(a: number | string) {
    return a;
  }
}

function test(a: number | string, c: testInterface) {
  const b = c.foo(a);
}
```
The first style is how Microsoft declare overloading methods in their projects, such as [Language Server Protocol](https://github.com/microsoft/vscode-languageserver-node/blob/1320922f95ef182df2cf76b7c96b1a2d3ba14c2a/client/src/common/features.ts#L664). However, it breaks `method-signature-style`. (To reproduce, click the typescript-eslint playgound link)

The second and third style is correct for `method-signature-style`, but it fails the type check when implementing those methods in a class. If I remove the overloading `number | string` signature, then it fails the type check when invoking method foo of the interface. (To reproduce, click the TS playgound link)

[TypeScript Playground Link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAkgZzASXNeTkDeAUMsgPRnIwD21AFHAFzIgCuAtgEbQCUzbXaAG4S5SjXpNkGKKADmfaWFkg5I0hSq0G-DtyjIAPkpULdgg8ZnyRozROZ0dLPb2QBeAHwuLPZADJkJylrVT8vE3k-QODzfUNQhQ9vAXjEnltSB0JRUlJnVN444Vy853TmRPU8-KlCy0iw4oaq0QBfIg6iBAAbODQ0ZAAVCAwAYT6B5GB2AAceiHYIcEH0LBxYRBRiLO061yhFeuqJcuUoyvPVE73mo0aknbyoCDBWKBBkOGqOrphWEAIMDAaifNYFA73RIAGmQCGYa2wkE2SD8TwQoIwyE4HjhADpTnAMp0gA)
[typescript-eslint Playground](https://typescript-eslint.io/play/#ts=5.0.2&jsx=true&sourceType=module&showAST=es&code=JYOwLgpgTgZghgYwgAkgZzASXNeTkDeAUMsgPRnIwD21AFHAFzIgCuAtgEbQCUzbXaAG4S5SjXpNkGKKADmfaWFkg5I0hSq0G-DtyjIAPkpULdgg8ZnyRozROZ0dLPb2QBeAHwuLPZADJkJylrVT8vE3k-QODzfUNQhQ9vAXjEnltSB0JRUlJnVN444Vy853TmRPU8-KlCy0iw4oaq0QBfIg6iBAAbODQ0ZAAVCAwAYT6B5GB2AAceiHYIcEH0LBxYRBRiLO061yhFeuqJcuUoyvPVE73mo0aknbyoCDBWKBBkOGqOrphWEAIMDAaifNYFA73RIAGmQCGYa2wkE2SD8TwQoIwyE4HjhADpTnAMp0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AW0XwAsB7AE1rJKAcyYBDfLGiIhRJOjABtSImjRe0SAF1wYAL4h9QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

Therefore, I need some suggestions:
Should I just disable `method-signature-style` for those parts, or open an issue in typescript-eslint, or open an issue in TypeScript?

**Comment by yf-yang** [comment]
Let me try to summarize: the rule `method-signature-style` means, when I am declaring an interface method, it is possible that someone implements with either `foo = () => {}` or `foo() {}`. For the first one, method style may break typescript strict function check.

So, if I can guarantee this interface is only implemented in classes whose methods are declared in the second style, then I can safely ignore `method-signature-style`. Therefore, I should disable this rule in such scenario.

Is it correct?

**Comment by bradzacher** [comment]
As per our contributing guidelines - please don't use discussions for bug reports or support requests.
https://typescript-eslint.io/contributing/discussions

Instead follow our issues and support workflows:
https://typescript-eslint.io/contributing/issues#questions-and-support-requests
